File: cpp_build_scan/cimage/cimage/ImgCollectionBuilder.cpp
/*
 * Copyright (C) 2018 russell brown
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#include "cimage.h"


 // ImgCollectionBuilder. This is a helper class to build an Image Collection.
 // It can use a simple or multithreaded approach.


// constructor. set the singleton
ImgCollectionBuilder::ImgCollectionBuilder(CreateType ct)
{
	createType = ct;
	//instance = this;
	ic = new ImgCollection();

	// Build & Search can be multithreaded, find out how many  based on the number of
	//  processors/cores in the machine. 
	
	if (createType == CREATETHREADS)
	{
		numThreads = std::thread::hardware_concurrency() * 2; // *2 seems to work well, more or less is slower
		if (numThreads <= 0)
			numThreads = 2;
	}
	else
		numThreads = 0;
}

void ImgCollectionBuilder::Save()
{
	ic->Save(saveTo);
}

// initCreate. only called from create, creates the image encoding threads
void ImgCollectionBuilder::initCreate()
{
	// flag used to end threads when needed
	running = true;

	// start each encoder if configured
	for (int t = 0; t < numThreads; t++)
	{
		RunThreadInfo *rti = new RunThreadInfo();
		rti->id = t;
		rti->trd = thread(&ImgCollectionBuilder::imgProcessingThread, this, rti);
		threads.push_back(rti);
		st.incThreads();
	}
	
}


// The 'image processing' thread
void ImgCollectionBuilder::imgProcessingThread(RunThreadInfo *ri)
{
	logger::info("Thread " + to_string(ri->id) + " running");
	ImageInfo *current = nullptr;
	int tp = 0;

	//
	// loop round reading items to process off the threadFeeder queue (populated by the fileWalker )
	//
	while (true)
	{
		// get next job
		tflock.lock();
		int ts = threadFeeder.size();
		if (ts > 0)
		{
			current = threadFeeder.front();
			threadFeeder.pop_front();
		}
		tflock.unlock();
		// nothing to do, have a nap and try again
		if (current == nullptr)
		{
			if (!running)
				break;
			MSNOOZE(10);
			continue;
		}

		tp++;
		logger::debug("Processing " + current->filepart + " in thread " + to_string(ri->id) );
		try
		{
			if ( ImgUtils::GetImageInfo(current) ) 
				processItemResult(current);
			else
				st.incErrors();
		}
		catch (std::runtime_error &e)
		{
			logger::error("Error processing " + current->filepart + ", " + string(e.what()));
		}
		catch (std::exception &e)
		{
			logger::error("Error processing " + current->filepart + ", " + string(e.what()));
		}
		current = nullptr;
	}
	logger::info("Thread " + to_string(ri->id) + " stopping, it processed " + to_string(tp) + " images.");
}


// Two ways to use the builder, either Create ( build from files ) or
// Load ( load an existing set ) 
// in either case the 'ic' will be filled
void ImgCollectionBuilder::Create(fs::path _top, fs::path path, fs::path _saveTo)
{
	saveTo = _saveTo;
	ic->top = _top;
	ic->stop = ic->top.string();
	ImgUtils::Replace(ic->stop, "\\", "/");

	initCreate();
	// call file walker with the 'top' directory
	walkFiles(path);
}

uint32_t ImgCollectionBuilder::pathsplit(const fs::path d, string &dirpart, string &filepart, time_t &lastMod)
{
	try
	{
		lastMod = 0;
		struct stat result;
		if (stat(d.string().c_str(), &result) == 0)
			lastMod = result.st_mtime;


		if (fs::is_regular_file(d))
		{
			filepart = d.filename().string();
			dirpart = d.parent_path().string();
		}
		else if (fs::is_directory(d))
		{
			filepart = "";
			dirpart = d.string();
		}
		else
			return 0;
	}
	catch (std::exception e)
	{
		logger::warn("Ignore ");
		return 0;
	}

	ImgUtils::Replace(dirpart, "\\", "/");
	dirpart = dirpart.substr(ic->top.string().length());

	if ( dirpart.empty() || dirpart[0] != '/' )
		dirpart = "/" + dirpart;

	return ImgUtils::GetHash(dirpart);
}

//
// walkFiles
//
// The main 'Create' method. Called initially from the 'top' directory
// we then loop through files in that dir, if we find a directory, call ourselves
// recursivly. If we find a file, call processFile to deal with it ( in 
// imgProcessingThread )
//
bool ImgCollectionBuilder::walkFiles(fs::path dir)
{

	string dirpart;
	string filepart;
	time_t lastmod = 0;

	// add 'top' directory to the collection 
	uint32_t dirHash = pathsplit(dir, dirpart, filepart, lastmod);
	ic->dirs.push_back(new ImgCollectionDirItem(dirHash, dirpart, lastmod));
	st.incDirs();

	// loop for all its files/subdirs
	int fcntr = 0;
	for (fs::directory_entry de : fs::recursive_directory_iterator(dir))
	{
		string dirpart;
		string filepart;

		// get crc32 of the directory path
		uint32_t dirHash = pathsplit(de, dirpart, filepart, lastmod);
		if (dirHash == 0)
		{
			st.incNameErrors();
			continue;
		}

		// givee occasional feedback
		if (++fcntr % 5000 == 0)
		{
			logger::info("At: " + st.progressStr());
		}

		// if its a directory, create a DirItem and add to collection
		if (filepart.empty())
		{
			st.incDirs();
			ic->dirs.push_back(new ImgCollectionDirItem(dirHash, dirpart, lastmod));
			logger::debug("Dir is " + dirpart);
		}
		else
		{
			// if its an image file, create a ImageInfo and process it
			st.incFiles();
			if (ImgUtils::IsImageFile(de.path()))
			{
				ImageInfo *ii = new ImageInfo();
				ii->de = de;
				ii->dirhash = dirHash;
				ii->filepart = filepart;
				processItem(ii);
			}
		}
	}

	// wait for image processing threads to stop
	waitOnProcessingThreads();

	return true;
}

void ImgCollectionBuilder::waitOnProcessingThreads()
{
	logger::info("Running set to false");
	running = false;

	if (createType == CREATETHREADS)
	{
		MSNOOZE(10); // allow any thread to pick up final task if set. there is a better way to do this... 
		for (list<RunThreadInfo*>::iterator rt = threads.begin(); rt != threads.end(); rt++)
		{
			(*rt)->trd.join();
			logger::debug("Finished " + to_string((*rt)->id) + " has finished normally.");
		}
	}


}
	
void ImgCollectionBuilder::processItemResult(ImageInfo *ii)
{
	//logger::info("Results " + ii->filepart);
	string err;

	st.addBytes(ii->size);
	{
		scoped_lock sl(llock);
		try
		{
			ic->files.push_back(new ImgCollectionFileItem(ii->dirhash, ii->crc, ii->filepart));
			if (ic->images.find(ii->crc) == ic->images.end())
			{
				ic->images[ii->crc] = new ImgCollectionImageItem(ii->crc, ii->thumb);
				st.incImages();
			}
			else
				st.incDuplicates();
		}
		catch (exception &e)
		{
			st.incErrors();
			err = "Could not gen image " + ii->filepart + string(e.what());
		}
	}

	delete ii;


	if (err.length() > 0)
		logger::error(err);

}

void ImgCollectionBuilder::processItem(ImageInfo *ii)
{
	// process an image file
	
	try
	{
		// if using threads, add the ImageInfo to the thread feeder queue, 
		// a thread will pick it off when it is ready
		if (createType == CREATETHREADS)
		{
			// if too many encodes in Q just wait
			while (threadFeeder.size() > (threads.size() * 10))
				MSNOOZE(10);

			// add to queue (protected as threads will access and remove items )
			tflock.lock();
			threadFeeder.push_back(ii);
			tflock.unlock();
		}
		else
		{
			//otherwise, just do what the thread would have done
			if (ImgUtils::GetImageInfo(ii))
				processItemResult(ii);
			else
				st.incErrors();
		}

	
	}
	catch (runtime_error &e)
	{
		logger::error("Could not gen image");
		printf("%s\n", e.what());
		printf("\n");
		st.incErrors();
	}
	catch (exception &e)
	{
		logger::error("Could not gen image");
		printf("%s\n", e.what());
		printf("\n");
		st.incErrors();
	}

}

//
// Save
//
// This saves the ImgCollection as three seperate files, dirss, files & images
//

File: jfind/src/scan/ImgCollectionBuilder.java
/*
 * Copyright (C) 2018 russell brown
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package scan;

import java.nio.file.Path;
import java.nio.file.SimpleFileVisitor;


// ImgCollectionBuilder. This is a helper class to build an Image Collection.
// it is derived from SimpleFileVisitor as one of its functions will be
// to scan all the image files in the 'top' directory
public class ImgCollectionBuilder extends SimpleFileVisitor<Path>
{
 /*
    public ImgCollectionBuilder(boolean ut)
    {
        useThreads = ut;
        set = new ImgCollection(ut);
    }
    // The collection we are building
    private ImgCollection set = null;

    // signal to stop threads
    public boolean running = true;
    // number of processing threads
    private int numThreads = 0;
    private boolean useThreads = true;
    // keep some stats on whats happening
    private final Stats st = new Stats();
 
    // ImgProcessingThread. This is the thread that will processs the images
    // in the scan. 
    class ImgProcessingThread implements Runnable
    {
        

        // Info about our thread
        public int id;
        public Thread thread;

        // the thread itself
        public void run()
        {
            // keep thread running, we will exit when running is set to false
            // but we cant do a while(running) as we may loose something in
            // our input q
            while (true)
            {
                // read an ImageInfo off our input queue
                ImageInfo ii = null;
                ii = threadFeeder.poll();

                // if something to process, do it...
                if (ii != null)
                {
                    processFile(ii);
                } else 
                // otherwise Q is empty, we can check running to see 
                // if we should exit the thread
                {
                    if (!running)
                    {
                        break;
                    }
                    // nothing to do, end not indicated. take a nap...
                    arenbee.jutils.Gen.Sleep(10);
                }
            }
            Logger.Info("** EXIT Thread " + id + " is exiting.");
        }
    }
    
    // used to pass info on an image file to the processing thread
    class ImageInfo
    {
        ImgCollectionFileItem sfile;
        Path name;
        Path dir;
    }

    // The main thread is driven by the file visitor mechanism. we 
    // feed an image file to one of the processing threads. 
    // we don't create a processing thread for each individual image as we 
    // would rapidly run out of resources. we need to limit this somehow. we
    // implement this by creating a processing thread for each processor 
    // on our host. These threads are added to the 'threads' list
    public List<ImgProcessingThread> threads = new LinkedList<>();
    // we feed images to the threads using the threadFeeder, put an image into the 
    // thread and led the individual threads remove one when it is ready. this
    // is thread safe
    public ConcurrentLinkedQueue<ImageInfo> threadFeeder = new ConcurrentLinkedQueue<>();

 

    private void setUpCreateThreads()
    {
        // create the processing threads. See how many processors we have
        Runtime rt = Runtime.getRuntime();
        numThreads = rt.availableProcessors() * 2;// - 1; // don't hog all threads
        // unless we only have the one processor, in which case we will have to hog it...
        if (numThreads < 1)
            numThreads = 1;
        
        Logger.Info("Running with " + numThreads + " threads");
        
        // create the threads and start them up. They will monitor the 
        // threadFeeded Q for things to do
        for (int tx = 0; tx < numThreads; tx++)
        {
            ImgProcessingThread rti = new ImgProcessingThread();
            rti.id = tx;
            rti.thread = new Thread(rti);
            rti.thread.start();
            threads.add(rti);
            st.incThreads();
            Logger.Info("Creating thread " + tx);
        }

    }
    
    // Two ways to use the builder, either Craete ( build from files ) or
    // Load ( load an existing set ) 
    // in either case the 'set' will be filled
    public void Create(Path spath, Path dpath)
    {
        
        // Make sure the dir to add exists
        if (!Files.exists(dpath))
        {
            Logger.Fatal("Directory " + dpath + " dosn't exist");
        }

        // If the set directory dosn't exist then make it
        if (!Files.exists(spath))
        {
            try
            {
                Files.createDirectories(spath);
            } catch (IOException ex)
            {
                Logger.Fatal("Set path " + spath + " dosn't exist and I couldn't create it.");
            }
        }

        // if creating, create the threads. we dont need to do this for a 'Load'
        if ( useThreads )
            setUpCreateThreads();

        try
        {
            // Set the top directory in the imageset 
            Timer.stagestart();
            set.SetTop(dpath);

            // And build....
            Files.walkFileTree(dpath, this);
            
            // All files & dis added, wait for any thumbnails to finish
            // encoder.WaitTillFinished();
            if ( useThreads )
                waitForThreads();
            
            // and save to disk
            Timer.stagestop("Create");
            set.Save(spath);

        } catch (IOException ex)
        {
            Logger.Severe("Error while scanning dir " + spath + " : " + ex.toString());
        }
        Logger.Info(st.toString());
    }    
    
    // Load. load from disk reather than scanning files
    public boolean Load(Path spath)
    {
        set.Load(spath);
        return true;
    }

    // called when the SimpleFileVisitor finds a directory
    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
    {
        st.incDirs();
        
        // add the dir
        set.AddDir(dir);
        return CONTINUE;
    }

    // called when the SimpleFileVisitor finds a file
    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
    {
        st.incFiles();

        Path name = file.getFileName();
        Path dir = file.getParent();
        boolean isImage = arenbee.jutils.Gen.IsImageFile(name.getFileName());
        if ( !isImage )
            return CONTINUE;

        // Find the diritem for this file
        ImgCollectionDirItem sid = set.GetDir(dir);
        if (sid == null)
        {
            Logger.Severe("No dir for a file!" + file);
            return CONTINUE;
        }

        // Create an imageinfo for this file 
        ImageInfo ii = new ImageInfo();
        // add the file to the image collection as a placeholder without the crc or  
        // thumb calculated, these will be added when the img processor thread
        // has processed the file
        ii.sfile = set.AddFile(sid, file);

        // send to a thread
        if ( useThreads )
            add(ii);
        else
            processFile(ii);
        
        // let user know we are still alive
        if (st.getFiles() % 50 == 0)
        {
            Logger.Info("At files=" + st.getFiles());
        }

        return CONTINUE;
    }


    void add(ImageInfo ii)
    {
        // sent to thread. If the threadFeeder is getting full
        // just wait until the threads catch up ( otherwise 
        // threadFeeder will just get huge
        while (threadFeeder.size() > numThreads * 2)
        {
            arenbee.jutils.Gen.Sleep(10);
        }
        threadFeeder.add(ii);
    }

    // processFile. This is where the main work is done! called from
    // an image processing thread
    void processFile(ImageInfo ii)
    {

        // create the imageitem
        ImgCollectionImageItem sii = null;
        Logger.Debug("processFile(1), name=" + ii.sfile);
        Path fpath = GetFullPathOfFile(ii.sfile);
        
        // get the thumb & crc of the data in the file
        if (Files.exists(fpath))
        {
            sii = getFileThumbAndCrc(fpath);
            if ( sii != null )
            {
                Logger.Debug("processFile, name=" + ii.sfile.getFile() + ", crc=" + Long.toHexString(sii.getIHash()));
                // call file processes to send resuts back to main thread
                FileProcessed(ii.sfile, sii);
                return;
            }
        }
        Logger.Warn("processFile, error processing name=" + ii.sfile.getFile() + ", path" + fpath.toString());
    }

    // Create the imageitem & fill the crc & thumb 
    public static ImgCollectionImageItem getFileThumbAndCrc(Path fpath)
    {
        ImgCollectionImageItem sii = new ImgCollectionImageItem();
        // call image helper to get the bits we want
        Image.SSCRC cby = Image.GetFilesImageComponents(fpath);
        if ( cby == null )
            return null;
        // sad save to the imageitem
        sii.setCrc(cby.crc);
        sii.setThumb(cby.thumb);
        sii.setSize(cby.bytes.length);
        return sii;
    }

    // GetFullPathOfFile. Gets the path of a fileitem using its dir hash to 
    // get the dir it is in and add the name
    public Path GetFullPathOfFile(ImgCollectionFileItem sf)
    {
        // full path is gotten from top + diritem realtive path + name
        ImgCollectionDirItem sid = set.GetDirByHash(sf.getDhash());
        String tpath = set.GetTop();
        String dpath = sid.getDir();
        String fpath = sf.getFile();

        String path = tpath + dpath + "/" + fpath;
        return Paths.get(path);
    }

    public FileVisitResult visitFileFailed(Path file, IOException exc)
    {
        st.incErrors();
        return CONTINUE;
    }

    // waitForThreads. 
    void waitForThreads()
    {
        // tell threads to stop
        running = false;
        
        // join them all before returning
        Logger.Info("Wait for encoder threads to finish...");
        for (ImgProcessingThread rti : threads)
        {
            try
            {
                rti.thread.join();
            } catch (InterruptedException ex)
            {
                Logger.Info(ex.getMessage());
            }
        }
        
        // stop timer
        st.stop();
    }


    // FileProcessed. This will be called by the encoder after a file has been processed. 
    void FileProcessed(ImgCollectionFileItem sif, ImgCollectionImageItem sii)
    {
        // add the image to the colelction. It may already be there from another
        // file
        boolean isNew = set.AddImage(sii);
        
        // inc stats
        st.addBytes(sii.getSize());
        if ( isNew )
            st.incImages();
        else
            st.incDuplicates();
        
        // record images crc map in fileitem
        sif.setIHash(sii.getIHash());
    }
*/
}
